keps/sig-auth/4412-projected-service-account-tokens-for-kubelet-image-credential-providers/README.md

```diff
@@ -0,0 +1,1107 @@
+proposal will be implemented, this is the place to discuss them.
+-->
+
+#### Credential Provider Configuration
+
+Adding a new field `TokenAttributes` to the `CredentialProvider` struct in the kubelet configuration.
+
+1. `ServiceAccountTokenAudience` is the intended audience for the credentials. If set, the kubelet will generate a service account token for the audience and pass it to the plugin.
+   1. Only a single audience can be configured.
+   2. The credential provider can be set up multiple times for each audience.
+   3. By setting this field, the credential provider is opting into using service account tokens for image pull.
+2. `ServiceAccountAnnotationKeys` is the list of annotation keys that the plugin is interested in. The keys defined in this list will be extracted from the corresponding service account
+   and passed to the plugin as part of `CredentialProviderRequest`. Plugins can use to extract additional information required to fetch credentials. The plugin is responsible for validating the existence of annotations and their values.
+   1. Service account annotations are significantly more stable (and map naturally to the service account being the partioning dimension).
+   2. This field makes the provider opt into getting the annotations it wants, avoids sending arbitrary annotation contents down to the plugin (including stuff like client-side apply annotations) and shrinks the set of annotations that could invalidate the cache.
+   3. This field can't be set without setting the `ServiceAccountTokenAudience` field.
+
+The configuration will not support a plugin to function in 2 modes simultaneously (one with service account tokens and one without). If a plugin needs to function in both modes, it will need to be configured twice with different names. This can be done to facilitate a gradual migration to the new mode for the plugin on a per-image or per-registry basis.
+
+```diff
+--- a/pkg/kubelet/apis/config/types.go
++++ b/pkg/kubelet/apis/config/types.go
+@@ -653,10 +653,19 @@ type CredentialProvider struct {
+        // +optional
+        Env []ExecEnvVar
+ 
++       // TokenAttributes is the configuration for the service account token that will be passed to the plugin.
++       // The credential provider opts in to using service account tokens for image pull by setting this field.
++       // When this field is set, kubelet will generate a service account token bound to the pod for which the
++       // image is being pulled and pass to the plugin as part of CredentialProviderRequest along with other
++       // attributes required by the plugin.
++       //
++       // The service account metadata and token attributes will be used as a dimension to cache
++       // the credentials in kubelet. The cache key is generated by combining the service account metadata
++       // (namespace, name, UID, and annotations map with keys defined in ServiceAccountTokenAttribute.serviceAccountAnnotationKeys).
```

Comment:
I assume the values associated with those keys will also be part of the cache key?

Reply:
that's correct! I had used "annotations map" to indicate but will clarify to make it more clear.